common/check_internal.h

```diff
@@ -10,6 +10,20 @@
 
 namespace Carbon::Internal {
 
+// Evaluates a condition in a CHECK. This diagnoses if the condition evaluates
+// to the constant `true` or `false`.
+[[clang::always_inline]] constexpr bool
+// Trailing GNU function attributes are incompatible with trailing return types.
+// Filed as https://github.com/llvm/llvm-project/issues/118697
+// NOLINTNEXTLINE(modernize-use-trailing-return-type)
+CheckCondition(bool condition)
+    __attribute__((diagnose_if(condition, "CHECK condition is always true",
+                               "error")))
+    __attribute__((diagnose_if(!condition, "CHECK condition is always false",
+                               "error"))) {
```

Comment:
IIUC, you're testing both ways to see if this is constant evaluated. Had you considered doing something with `std::is_constant_evaluated`?

Reply:
(does is_constant_evaluated just not work here?)

Reply:
`is_constant_evaluated` doesn't help us here -- the call isn't manifestly constant evaluated, so it will always return false. There are ways to do this detection that don't involve a Clang extension, but they're substantially more complex and would require that we parse and type-check each condition twice; I think `diagnose_if` is the cleanest mechanism available.